### PR TITLE
feat: make publishing event clearer

### LINKED ISSUE #4673: Wizard: Make Event Saving, Draft Mode, Publishing Clearer
**1. When event is created and still unpublished (Draft Mode)**
* _Requirement_ for saving a draft is only
  1. to have an event name
* show buttons:
  * Discard (only when event was never saved - when creating the event: It does not save the event and discards any data that was entered. After that direct user back to dashboard of all events. If user has saved an event previously, there is no "Discard" possible anymore, then show "Cancel" and send user back to specific event dashboard.)
  * Cancel (cancels changes of the wizard step and exists to event dashboard)
  * Save/Previous (except for first step)
  * Save/Next (saves changes and goes to the next step, except for last step)
  * Save (saves draft and exits wizard to dashboard)
* below the above option show another section with two buttons
  * Show a horizontal line with 100% width of the page
  * Show the text "This event is currently not published. It is in draft mode and it is not visible publicly."
  * Preview (open in new tab)
  * Publish (opens a pop up - same as "Publish" on the dashboard / afterwards exits to event dashboard). Please show this button as "gray" as long as the event does not have the minimum data listed necessary for publishing an event.

Screenshot from 2020-07-30 18-16-43

**2. After an event is published already (Live Event)**
* _Requirement_ for an event to be publishable is, that it has
  1. A name
  2. A location or online event link
  3. A ticket
* below the above savings/next/previous options show another section with two buttons
  * Show the text "This event is published. Any changes you make will appear on your live event."
  * View (open in new tab)
  * Unpublish (opens a pop up - same as "Unpublish" on the dashboard / afterwards exits to event dashboard)

Screenshot from 2020-07-30 18-23-19

**3. Top menu "Wizard Steps"**

When a user clicks a top wizard menu item the following action should take place:
* Cancel changes of the current tab and load the menu tab

Screenshot from 2020-07-30 19-24-21

### PR DESCRIPTION
**Pull request by maze-runnar**
Fixes #4673

from https://github.com/fossasia/open-event-frontend/pull/4942

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by maze-runnar** [comment]
@iamareebjamal please review

File: app/components/forms/wizard/sessions-speakers-step.js [hunk 1]
```diff
@@ -181,6 +185,13 @@ export default Component.extend(EventWizardMixin, FormMixin, {
         this.sendAction('save');
       });
     },
+    unpublish() {
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
+        this.sendAction('save', this.data);
+    });
```
Comment (niranjan94) [inline app/components/forms/wizard/sessions-speakers-step.js:192]: Codacy found an issue: [Expected indentation of 6 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/components/forms/wizard/basic-details-step.js [hunk 2]
```diff
@@ -497,4 +496,4 @@ export default Component.extend(FormMixin, EventWizardMixin, {
       this.onValid(() => {});
     }
   }
-});
+});
```
Comment (niranjan94) [inline app/components/forms/wizard/basic-details-step.js:499]: Codacy found an issue: [Newline required at end of file but not found.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/other-details-step.js [hunk 3]
```diff
@@ -164,6 +168,12 @@ export default Component.extend(FormMixin, EventWizardMixin, {
         this.sendAction('save');
       });
     },
+    unpublish() {
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
+        this.sendAction('save');
+     });
```
Comment (niranjan94) [inline app/components/forms/wizard/other-details-step.js:175]: Codacy found an issue: [Expected indentation of 6 spaces but found 5.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/sponsors-step.js [hunk 4]
```diff
@@ -68,6 +71,13 @@ export default Component.extend(FormMixin, {
         this.set('data.event.state', 'published');
         this.sendAction('save');
       });
+    },
+
+    unpublish() {
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
+        this.sendAction('save', this.data);
+       });
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:80]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Review by mariobehling — CHANGES_REQUESTED** [review]
Please read the issue description again carefully and rename the buttons accordingly. There is no point to review your PRs if you dont read the issues.



**1. When event is created and still unpublished (Draft Mode)**
* _Requirement_ for saving a draft is only
  1. to have an event name
* show buttons:
  * [x] Discard (only when event was never saved - when creating the event: It does not save the event and discards any data that was entered. After that direct user back to dashboard of all events. If user has saved an event previously, there is no "Discard" possible anymore, then show "Cancel" and send user back to specific event dashboard.)
  * [x] Cancel (cancels changes of the wizard step and exists to event dashboard)
  * [x] Previous (except for first step)
  * [x] Next (saves changes and goes to the next step, except for last step)
  * [x] Save draft (saves draft and exits wizard to this event's dashboard)
* [x] below the above option show another section with two buttons
  * [x] Show a horizontal line with 100% width of the page
  * [x] Show the text "This event is currently not published. It is in draft mode and it is not visible publicly."
  * [x] Preview (open in new tab)
  * [x] Publish (opens a pop up - same as "Publish" on the dashboard / afterwards exits to event dashboard). Please show this button as "gray" as long as the event does not have the minimum date listed necessary for publishing an event.

**2. After an event is published already (Live Event)**
* _Requirement_ for an event to be publishable is, that it has
  1. A name
  2. A location or online event link
  3. A ticket
* [x] below the above savings/next/previous options show another section with two buttons
  * [x] Show the text "This event is published. Any changes you make will appear on your live event."
  * [x] View (open in new tab)
  * [x] Unpublish (opens a pop up - same as "Unpublish" on the dashboard / afterwards exits to event dashboard)

**3. Top menu "Wizard Steps"**

When a user clicks a top wizard menu item the following action should take place:
* [ ] Cancel changes of the current tab and load the menu tab

**Comment by mariobehling** [comment]
Screenshot from 2020-09-08 12-47-34

Please also increase the spacing between the the first row of buttons and the second row "View/Publish" button.

As you can see in the screenshot the buttons show "Preview/Unpublish". Expected is "View/Unpublish". The View button does not open in a new tab as described in the issue.

Sorry, but this seems a bit unfair - do you expect other devs to review your PRs and point out each of the list items on the issue to you one by one?

Please go through each point first by yourself and do help us to shorten the review time that is required by paying more attention to each point. Thank you.

**Comment by maze-runnar** [comment]
Button Spacing, open preview in new window, adding horizontal line, Showing the text "This event is published. Any changes you make will appear on your live event." and "This event is currently not published. It is in draft mode and it is not visible publicly.".
Screenshot from 2020-09-08 21-10-35
Screenshot from 2020-09-08 20-49-28
Screenshot from 2020-09-08 20-39-27
Screenshot from 2020-09-08 19-51-26
Screenshot from 2020-09-08 19-51-21
Screenshot from 2020-09-08 18-57-33
Screenshot from 2020-09-08 18-57-28

**Comment by maze-runnar** [comment]
Publish and unpublish popup are not working the same as in the dashboard section. So made all other changes.
And making this final change - 
When a user clicks a top wizard menu item the following action should take place:
 Cancel changes of the current tab and load the menu tab

**Comment by mariobehling** [comment]
Thank you, please also change
* Rename "Forward" to -> "Next"
* Live events are missing the "Save" feature (Clicking "Save" should save the current wizard step and exit to the dashboard of this event)

Screenshot from 2020-09-08 18-40-42

**Comment by mariobehling** [comment]
When creating a new event, it is not clear yet how to publish it.

Screenshot from 2020-09-08 18-56-04

I think it should only be possible to publish an event once the event draft has been saved.

Therefore please display:
* A "Publish" button that is in grey color as long as the event has not been saved as a draft.
* A horizontal line as in other views
* In this initial step - before the event has been saved, please also show the following text above the horizontal line: "You can preview the event after you have saved it as a draft. You can publish the event after filling in the required fields."

**Comment by maze-runnar** [comment]
> * Live events are missing the "Save" feature (Clicking "Save" should save the current wizard step and exit to the dashboard of this event)

I am not getting it, why live events should option to save,while it can be published.

**Comment by maze-runnar** [comment]
> I think it should only be possible to publish an event once the event draft has been saved.
> 
> Therefore please display:
> A "Publish" button that is in grey color as long as the event has not been saved as a draft.
> 

previously it was mentioned in issue like toicket, location and name is required for publishing an event. After these changes user can't publish events in first time?

**Comment by maze-runnar** [comment]
changing msg when event is not saved for first time, and forward -> Next
Screenshot from 2020-09-08 23-54-45
Screenshot from 2020-09-08 23-37-28

**Comment by maze-runnar** [comment]
make publish button grey when event is not saved for first time - 
Screenshot from 2020-09-09 00-09-47

**Comment by maze-runnar** [comment]
Made required changes, not sure that save button should display when event is published.

**Comment by mariobehling** [comment]
> Made all the required changes, not sure that save button should display when event is published.

Yes, please display a "Save" button when the event is published as well.

**Comment by mariobehling** [comment]
Still missing:
* [ ] Show "Save" button after event is published
* [ ] Publish (opens a pop up - same as "Publish" on the dashboard / afterwards exits to event dashboard). 
* [ ] Unpublish (opens a pop up - same as "Unpublish" on the dashboard / afterwards exits to event dashboard)

Screenshot from 2020-09-08 20-53-28

**Comment by mariobehling** [comment]
Thanks for adding the "Save" button. Please change it in a way that it saves the event in the current "published" state. It currently unpublishes published events. Expected:
* unpublished events show "Save as draft"
* published events show "Save" (and keep the state published when saved)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by maze-runnar** [comment]
Completed PR, with Publish/Unpblish popup and Save/Save as Draft button implementation like mentioned 
Screenshot from 2020-09-09 10-46-45
Screenshot from 2020-09-09 10-46-23
Screenshot from 2020-09-09 10-46-07

**Comment by mariobehling** [comment]
Thank you, it looks great. But: the "Save" buttton still saves + unpublishes the event.

It should not unpublish an event.

**Comment by maze-runnar** [comment]
> Thank you, it looks great. But: the "Save" buttton still saves + unpublishes the event.
> 
> It should not unpublish an event.

ok, fixing it :+1:

**Comment by maze-runnar** [comment]
> Thank you, it looks great. But: the "Save" buttton still saves + unpublishes the event.
> 
> It should not unpublish an event.

Done

File: app/mixins/event-wizard.js [hunk 5]
```diff
@@ -210,6 +210,11 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
         this.sendAction('save');
       });
     },
+    savePublished() {
+      this.onValid(() => {
+        this.sendAction('save');
+      })
```
Comment (niranjan94) [inline app/mixins/event-wizard.js:216]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/components/forms/wizard/attendee-step.js [hunk 6]
```diff
@@ -23,16 +27,54 @@ export default Component.extend(FormMixin, {
         this.sendAction('save', this.data);
       });
     },
+    savePublished() {
+      this.onValid(() => {
+        this.set('data.event.state', 'published');
+        this.sendAction('save', this.data);
+      })
```
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:34]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/sponsors-step.js [hunk 7]
```diff
@@ -58,16 +61,53 @@ export default Component.extend(FormMixin, {
         this.sendAction('save');
       });
     },
+    savePublished() {
+      this.onValid(() => {
+        this.set('data.event.state', 'published');
+        this.sendAction('save', this.data);
+      })
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:68]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by iamareebjamal** [comment]
@maze-runnar While Mario checks if it works as expected, please pick up issues from invoice milestone

**Comment by mariobehling** [comment]
Thanks. Save button works as expected now. BUT: Unpublish button disappeared.

Screenshot from 2020-09-09 12-06-45

**Comment by maze-runnar** [comment]
@mariobehling , Unpublish button is corretly displaying on my local and even on deployment URL - 
Screenshot from 2020-09-09 15-44-57

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by maze-runnar** [comment]
@iamareebjamal why build is failing, there is nothing in logs..

**Comment by mariobehling** [comment]
> @mariobehling , Unpublish button is corretly displaying on my local and even on deployment URL -
> Screenshot from 2020-09-09 15-44-57

@iamareebjamal Is this related to existing events? No "Unpublish" button e.g. check here https://open-event-frontend-git-fork-maze-runnar-clearer.eventyay.vercel.app/events/5ac8e4d9/edit/basic-details

**Comment by maze-runnar** [comment]
> > @mariobehling , Unpublish button is corretly displaying on my local and even on deployment URL -
> > Screenshot from 2020-09-09 15-44-57
> 
> @iamareebjamal Is this related to existing events? No "Unpublish" button e.g. check here https://open-event-frontend-git-fork-maze-runnar-clearer.eventyay.vercel.app/events/5ac8e4d9/edit/basic-details

the unpublish button will not show up when event name, location, or ticket is missing like publish button.

**Comment by iamareebjamal** [comment]
@maze-runnar Good that you rebased the changes. We have changed publishability criteria to just event.name and tickets. Location is not required now after integration of online events.

**Comment by iamareebjamal** [comment]
Please check the changes here and make appropriate edits - https://github.com/fossasia/open-event-frontend/pull/5012

**Comment by iamareebjamal** [comment]
Sometimes build fails when heroku link goes to sleep. Restarted

File: app/components/forms/wizard/attendee-step.js [hunk 8]
```diff
@@ -23,16 +27,54 @@ export default Component.extend(FormMixin, {
...
-        this.sendAction('save', this.data);
+        this.sendAction('move', direction, this.data);
       });
+    },
+
+    openConfirmModal() {
+      this.set('isPublishUnpublishModalOpen', true);
+    },
+
+    togglePublishState() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/attendee-step.js:46]: Please move all common code to a shared instance after the changes are functionally approved. Same block of code should not be repeated anywhere

**Comment by iamareebjamal** [comment]
See this is why we want to avoid duplication. Same logic needed to be modified in 4 places separately. We don't want to repeat the effort like this

**Comment by maze-runnar** [comment]
> See this is why we want to avoid duplication. Same logic needed to be modified in 4 places separately. We don't want to repeat the effort like this

once the working is confirmed I will remove duplication.

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/components/forms/wizard/attendee-step.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by maze-runnar** [comment]
adding publish modal function to event wizard and inheriting them works in first two steps but in rest it give error and gives a blank page - 
Screenshot from 2020-09-09 17-38-54

**Comment by maze-runnar** [comment]
Can I work on other issues now, if this is completed. I think this is working as expected by @mariobehling explanation.

**Comment by iamareebjamal** [comment]
Yes, please solve other issues meanwhile

**Review by mariobehling — APPROVED** [review]
Working as expected in my tests. Thanks.

**Comment by iamareebjamal** [comment]
Great. @maze-runnar Please deduplicate the code tomorrow as much as you can first and then I'll polish it

File: app/components/forms/wizard/other-details-step.js [hunk 9]
```diff
@@ -28,6 +28,10 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return this.data.event.socialLinks.filterBy('isDeleted', false);
   }),
 
+  ticketsPresent: computed('data.event.tickets.@each', function() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/other-details-step.js:31]: This is still duplicated

File: app/components/forms/wizard/sponsors-step.js [hunk 10]
```diff
@@ -58,16 +61,53 @@ export default Component.extend(FormMixin, {
...
-        this.set('data.event.state', 'published');
-        this.sendAction('save');
+        this.sendAction('move', direction);
       });
+    },
+    openConfirmModal() {
+      this.set('isPublishUnpublishModalOpen', true);
+    },
+
+    togglePublishState() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/sponsors-step.js:79]: This is still repeated

**Comment by iamareebjamal** [comment]
An action and some computed properties are still duplicated

File: app/components/forms/wizard/sponsors-step.js (hunk 10 above)
Comment (maze-runnar) [reply app/components/forms/wizard/sponsors-step.js:79]: sponsor and attendee can't inherit EventWizard

Comment (maze-runnar) [reply app/components/forms/wizard/sponsors-step.js:79]: only basis-step-detail, other-details, session-speaker can inherit. On inheriting in sponsors-step.js, And attendee it is causing error and other functionalities are not working as expected.

**Comment by maze-runnar** [comment]
> An action and some computed properties are still duplicated

they can be only deduplicated for basic, other, and session steps.

File: app/components/forms/wizard/sponsors-step.js (hunk 10 above)
Comment (iamareebjamal) [reply app/components/forms/wizard/sponsors-step.js:79]: You can create another mixin and inherit that in those steps. There is always a way to remove duplication

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/components/forms/wizard/attendee-step.js [hunk 11]
```diff
@@ -1,8 +1,9 @@
 import Component from '@ember/component';
 import { computed } from '@ember/object';
 import FormMixin from 'open-event-frontend/mixins/form';
+import EventWizardMixin from 'open-event-frontend/mixins/event-wizard';
 
-export default Component.extend(FormMixin, {
+export default Component.extend(FormMixin,EventWizardMixin, {
```
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/components/forms/wizard/attendee-step.js (hunk 11 above)
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/sponsors-step.js [hunk 12]
```diff
@@ -1,8 +1,9 @@
 import Component from '@ember/component';
 import { computed } from '@ember/object';
 import FormMixin from 'open-event-frontend/mixins/form';
+import EventWizardMixin from 'open-event-frontend/mixins/event-wizard';
 
-export default Component.extend(FormMixin, {
+export default Component.extend(FormMixin,EventWizardMixin, {
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by iamareebjamal** [comment]
`preSaveActions` is not called in all necessary places. Check the code changes and call where necessary

**Comment by maze-runnar** [comment]
> `preSaveActions` is not called in all necessary places. Check the code changes and call where necessary

I should remove actions that are not in use?

**Comment by iamareebjamal** [comment]
See the changes you made. It's a function I added in https://github.com/fossasia/open-event-frontend/pull/5012

**Comment by iamareebjamal** [comment]
I'm curious as to why you changed the method `publish` name to `savePublish`. As far as I can see, it is exactly same as publish function. We need to make as small amount of changes as possible. Now, why not just keep the publish method as is? Or is there some reason

File: app/components/forms/wizard/attendee-step.js (hunk 11 above)
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/mixins/event-wizard.js [hunk 13]
```diff
@@ -45,6 +46,10 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
       }
     ];
   },
+
```
Comment (niranjan94) [inline app/mixins/event-wizard.js:49]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/sponsors-step.js (hunk 12 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by maze-runnar** [comment]
> I'm curious as to why you changed the method `publish` name to `savePublish`. As far as I can see, it is exactly same as publish function. We need to make as small amount of changes as possible. Now, why not just keep the publish method as is? Or is there some reason

All the publish/unpublish code is moved to event-wizard. It is saving the event when it is already publish. so avoid the confusion I renamed it to savePublish.

**Comment by iamareebjamal** [comment]
I'll be polishing the PR and reverting the code changes. Try to make as few internal code changes in future. If the function code doesn't change, the name shouldn't change. At the end of the day, it is still publishing the event.

File: app/mixins/event-wizard.js [hunk 14]
```diff
@@ -210,6 +216,12 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
         this.sendAction('save');
       });
     },
+    savePublished() {
+      this.onValid(() => {
+        preSaveActions.call(this);
```
Comment (iamareebjamal) [inline app/mixins/event-wizard.js:221]: Here savePublished does not set state to be published. Hence I understand, this is why is is savePublished

File: app/components/forms/wizard/attendee-step.js [hunk 15]
```diff
@@ -23,15 +24,15 @@ export default Component.extend(FormMixin, {
         this.sendAction('save', this.data);
       });
     },
-    move(direction) {
+    savePublished() {
       this.onValid(() => {
-        this.sendAction('move', direction, this.data);
+        this.set('data.event.state', 'published');
```
Comment (iamareebjamal) [inline app/components/forms/wizard/attendee-step.js:29]: Here savePublished sets the event to be published, so why not just use publish method

**Comment by iamareebjamal** [comment]
I'm still confused. There are subtle differences in the overridden methods for some reason. Can you please check the latest comments? It makes reasoning very difficult

**Comment by maze-runnar** [comment]
Savepublish will not set state to publish, I am removing that line.

**Comment by iamareebjamal** [comment]
Now this makes a lot more sense. But, since `savePublished` and `move` are already defined in event wizard mixin, they shouldn't be needed in the individual components, right?

**Comment by maze-runnar** [comment]
> Now this makes a lot more sense. But, since `savePublished` and `move` are already defined in event wizard mixin, they shouldn't be needed in the individual components, right?

the move is needed due to some direction in action, and yes savePublished is not needed , removing that too.

**Comment by maze-runnar** [comment]
onMoving  savePublished to event-wizard for attendee it's giving error - "can't read property custom form of undefined."

File: app/mixins/event-wizard.js (hunk 13 above)
Comment (niranjan94) [inline app/mixins/event-wizard.js:49]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/components/forms/wizard/attendee-step.js (hunk 11 above)
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

File: app/components/forms/wizard/sponsors-step.js (hunk 12 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:6]: Codacy found an issue: [A space is required after ','.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894)

**Comment by iamareebjamal** [comment]
OK. Thanks

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/controllers/events/view.js  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by iamareebjamal** [comment]
@maze-runnar Please check this commit - [`927b3ce` (#5011)](https://github.com/fossasia/open-event-frontend/pull/5011/commits/927b3ce68ab61379fe042c8daf00bfbb35c4543e)

This does not contain any duplicated code. Nor the View/Preview Publish/Unpublish code nor the Next/Previous code. So, in future, please move features into composable components and reuse them across the project. Copy Pasting is easy but very error prone and hard to maintain/review. For example, in 2nd step, there are still previous UI remaining. Secondly, the publishing on wizard page did not highlight the missing form fields in ticket form (now fixed)

There is still one issue remaining. It was there before my changes and after as well. Previously, whenever we pressed `Enter` on any field, it submitted the form and moved to next. But now, it does that but also opens the unpublishing dialog and dims the screen. Please check why does this happen and try to fix it

**Comment by maze-runnar** [comment]
now it's working correctly

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/controllers/events/view.js  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/mixins/event-wizard.js [hunk 16]
```diff
@@ -46,6 +47,10 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
     ];
   },
 
+  ticketsPresent: computed('data.event.tickets.@each', function() {
```
Comment (iamareebjamal) [inline app/mixins/event-wizard.js:50]: Why bring this back? It's not needed

File: app/mixins/event-wizard.js [hunk 17]
```diff
@@ -227,6 +232,44 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
         callback(true);
       });
     },
+
+    openConfirmModal() {
```
Comment (iamareebjamal) [inline app/mixins/event-wizard.js:236]: Why bring this back? It's not needed

File: app/mixins/event-wizard.js [hunk 18]
```diff
@@ -227,6 +232,44 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
...
+      this.set('isPublishUnpublishModalOpen', true);
+    },
+
+    togglePublishState() {
```
Comment (iamareebjamal) [inline app/mixins/event-wizard.js:240]: Why bring this back? It's not needed

**Comment by iamareebjamal** [comment]
You have added some functions back which are not needed. Please remove them

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/controllers/events/view.js  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/mixins/event-wizard.js [hunk 19]
```diff
@@ -232,54 +227,11 @@ export default Mixin.create(MutableArray, CustomFormMixin, {
...
-        .finally(() => {
-          this.set('isLoading', false);
-        });
-    },
     addItem(type, model) {
       if (type === 'socialLinks') {
         this.get(`data.event.${type}`).pushObject(this.store.createRecord(model, {
           identifier: v1()
         }));
-      } else if (type === 'customLink') {
```
Comment (iamareebjamal) [inline app/mixins/event-wizard.js:231]: Why remove this?

**Comment by iamareebjamal** [comment]
Um, now you removed a random if statement which was necessary. There always seems to be either some random think added or removed. Please carefully review the changes yourself before committing. There are way more review cycles than needed. You need to improve the self-evaluation phase of your PRs

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/controllers/events/view.js  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

File: app/mixins/event-wizard.js (hunk 16 above)
Comment (iamareebjamal) [reply app/mixins/event-wizard.js:50]: Now this is back again

**Comment by iamareebjamal** [comment]
Now `ticketsPresent` is back again. Please review the changes locally before committing and then before pushing

**Comment by maze-runnar** [comment]
> Um, now you removed a random if statement which was necessary. There always seems to be either some random think added or removed. 

yeah, I don't know why this happens to me a lot, and in this PR more than a lot. I am trying to improve it.

**Comment by iamareebjamal** [comment]
Before pushing. Go through the code where you made the changes and check all the things previously mentioned are removed or not, and nothing else is removed. If you're trying to do it using GitHub UI, then it definitely won't be correct because it is extremely easy to make mistakes there and not realize what you are removing and what extra thing you removed. Nor is there any way to review the final changes before pushing

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/controllers/events/view.js  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6140894&bid=20076052)

**Comment by maze-runnar** [comment]
I think it's ok now.

**Review by iamareebjamal — APPROVED** [review]

**Review by iamareebjamal — APPROVED** [review]

**Comment by mariobehling** [comment]
Glad, it is working now! Thank you.

**Pull request merged**